### DISCUSSION TITLE
"Warning: the ECDSA host key for 'github.com' differs from the key for the IP address" issue

### DISCUSSION THREAD

**Discussion by liuzz10** [community/community#54604] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I don't know what happened, but when I got up and `git push` my changes it gave:
```
Warning: the ECDSA host key for 'github.com' differs from the key for the IP address '140.82.112.4'

Offending key for IP in /Users/joyliu/.ssh/known_hosts:13

Matching host key in /Users/joyliu/.ssh/known_hosts:16
```
It seems like I'm not able to connect to Github. I Googled the fix and did `sed -i '13d' ~/.ssh/known_hosts` to delete line 13 but then it shows the same error with a different line number.

It also happened when I did `git pull`.

When I tried VScode to sync changes it gave:
```
Host key verification failed.

fatal: Could not read from remote repository.
```
Can anyone help me? Thank you in advance!

**Comment by hrishikesh-kocharekar-dev** [comment]
Hi Liu, 
This error usually occurs when the host key for GitHub has changed, which can happen for a variety of reasons such as server maintenance or security updates. To fix this, you need to update your known_hosts file with the new host key for GitHub.

You can do this by running the following command in your terminal:

`ssh-keygen -R github.com`

This will remove the old key for GitHub from your known_hosts file. Then try to run the git command again, and it should prompt you to add the new host key to your known_hosts file. Just type "yes" to add the new key, and you should be able to connect to GitHub again.

**Reply by airtower-luna** [reply]
> Just type "yes" to add the new key, and you should be able to connect to GitHub again.

You should NEVER just type yes. You need to check first if the fingerprint you see is actually one of the [published GitHub key fingerprints](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints). If you don't do that, an attacker could intercept your connection. :warning: 

The same applies for any other SSH connection there'll just be different ways to get the right fingerprints.

**Reply by liuzz10** [reply]
Hi friend, thank you for helping me! It seems get updated successfully with the command. It gave:
```
# Host github.com found: line 15
# Host github.com found: line 18
# Host github.com found: line 19
# Host github.com found: line 20
/Users/joyliu/.ssh/known_hosts updated.
Original contents retained as /Users/joyliu/.ssh/known_hosts.old
```

But when I `git push`, it gave:
```
The authenticity of host 'github.com (140.82.112.3)' can't be established.
ECDSA key fingerprint is SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'github.com' (ECDSA) to the list of known hosts.
Warning: the ECDSA host key for 'github.com' differs from the key for the IP address '140.82.112.3'
Offending key for IP in /Users/joyliu/.ssh/known_hosts:12
Are you sure you want to continue connecting (yes/no)? yes
To github.com:liuzz10/liuzz10.github.io.git
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'github.com:liuzz10/liuzz10.github.io.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

When I did `git pull`, it gave:
```
Warning: the ECDSA host key for 'github.com' differs from the key for the IP address '140.82.114.3'
Offending key for IP in /Users/joyliu/.ssh/known_hosts:16
Matching host key in /Users/joyliu/.ssh/known_hosts:17
Are you sure you want to continue connecting (yes/no)? yes
fatal: refusing to merge unrelated histories
```
Do you know what's going on here? Thank you!

**Reply by airtower-luna** [reply, marked as answer]
Looks like SSH has been recording keys used for IP GitHub addresses, too, and now there's a conflict. You can delete the records for IP addresses the same way as for hostnames, e.g.:

```
ssh-keygen -R 140.82.114.3
```

GitHub has a bunch of IP addresses, so you might have to delete keys for multiple ones (all your SSH client has encountered before the key or key type changed).

**Reply by liuzz10** [reply]
Let me try! Do I need to re add key using this command `ssh-keyscan github.com >> ~/.ssh/known_hosts`?

**Reply by airtower-luna** [reply]
You should NEVER use that command: It adds keys without checking if they are correct, you wouldn't know if someone is manipulating your connection. If you already did, remove the keys it added (`ssh-keygen -R github.com`). :warning: 

Instead when SSH asks about an unknown key you need to *check if its actually the right one*. For GitHub the key fingerprints (to review) and keys you can copy & paste to `~/.ssh/known_hosts` (so the keys won't be unknown in the first place) are published [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints).

**Reply by liuzz10** [reply]
Thank you! I removed it. I also removed the 140.82.114.3 with this command `ssh-keygen -R 140.82.114.3`. But there came another IP 140.82.112.3. Should I remove this IP as well?

Here's the log:
```
(base) ➜  liuzz10.github.io git:(master) ssh-keygen -R 140.82.114.3
# Host 140.82.114.3 found: line 16
/Users/joyliu/.ssh/known_hosts updated.
Original contents retained as /Users/joyliu/.ssh/known_hosts.old
(base) ➜  liuzz10.github.io git:(master) git pull
The authenticity of host 'github.com (140.82.112.3)' can't be established.
ECDSA key fingerprint is SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'github.com' (ECDSA) to the list of known hosts.
Warning: the ECDSA host key for 'github.com' differs from the key for the IP address '140.82.112.3'
Offending key for IP in /Users/joyliu/.ssh/known_hosts:12
Are you sure you want to continue connecting (yes/no)? yes
fatal: refusing to merge unrelated histories
```

**Reply by airtower-luna** [reply]
> Should I remove this IP as well?

Yes, that's why I pointed out above that GitHub uses multiple IP addresses. Depending on which ones you happen to connect to (and which ones your computer already knew) you might need to do a few rounds of it. :sweat_smile:

You do *not* need to (and shouldn't) remove the `github.com` keys again.

**Reply by liuzz10** [reply]
It solved!!! Thank you for being so knowledgable and warm-hearted! I'm still confused about how it works and need to digest it a little bit. But now just want to say thank you!!!

**Reply by daniel-santos** [reply]
> Hi Liu, This error usually occurs when the host key for GitHub has changed, which can happen for a variety of reasons such as server maintenance or security updates. 

Somebody kindly correct me if I'm wrong, but isn't the only reason you should ever, *ever*, **EVER**, change your ssh host key is when you suspect that it has been compromised? As a 30 year software engineer veteran, to the very best of my understanding, there is absolutely NO other reason to change it. (Even if you want to move from one type of key to another? not sure how you do that properly). This isn't a specific area of my expertise.

Was their network compromised at some point that forced this perhaps?

EDIT: Oh! At least my understanding is solid, whew! It was for this reason: https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/

**Reply by airtower-luna** [reply]
Yeah, as long as the key hasn't been compromised *and* the type/size combination is still sound there's no reason to change it. If you want to do a smooth rollover, OpenSSH's `sshd` can announce additional keys to clients (all that are configured with `HostKey` in the server config), and the clients can then choose to add them to their `known_hosts` file (see [`UpdateHostKeys`](https://manpages.debian.org/bullseye/openssh-client/ssh_config.5.en.html#UpdateHostKeys)).

**Reply by bc1121** [reply]
After `ssh-keygen -R 140.82.114.3` (for my current offending IP address) I had to use this command several more times for other IP addresses (as mentioned above).

I didn't want security messages disrupting my flow for each ip address that needed to be removed when randomly encountered during clones, pushes, etc..., so I manually searched and deleted the remaining ssh-rsa keys from my known_hosts file.

To solve the problem once and for all, I ran diff on my known_hosts and known_hosts.old file to isolate the offending rsa key. Then I ran a search and highlight in VIM and used visual mode to delete all remaining entries ending in "= ssh-rsa {the offending key}"

**Comment by nvvetal** [comment]
Use this official instruction and it will solve your issue

https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/

**Reply by IRCraziestTaxi** [reply]
I'm getting a different fingerprint than the one that article specifies to check for. I'm not sure what to do.

**Reply by airtower-luna** [reply]
The key in that article is only the RSA key. Depending on your SSH config you may see a different one (Ed25519 or ECDSA). Depending on the type the fingerprint must be one of those listed in the documentation: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints

If it isn't one of those, *something* is manipulating your connection.

**Reply by mbesida** [reply]
The solution proposed in thig blog doesn't work. I'm still getting the warning after each pull/push. And each time the IP address is different
